py/foundation/geometry.py

```diff
@@ -5,242 +5,390 @@
 from math import sqrt
 from typing import FrozenSet, Generator, Iterable, Optional
 
+import numpy as np
```

Comment:
Hmmm that's an excellent point. I can probably refactor it out, just at the cost of some slight efficiency. But I doubt this will be run enough times for that to have a non-negligible effect. I'll change it.